src/pathfinding/dijkstra.ts

```diff
@@ -0,0 +1,80 @@
+import { Vector2 } from "../util";
+import * as Math from "../math";
+import { getRing4, getRing8 } from "../fov/get-ring";
+
+interface dijkstraConfig {
+  isBlockedCallback?: IsBlockedCallback;
+  topology: "four" | "eight";
+}
+
+type IsBlockedCallback = (v: Vector2) => boolean;
+
+export class Dijkstra {
+  readonly topology: "four" | "eight";
+  private readonly isBlocked?: IsBlockedCallback;
+
+  constructor(config: dijkstraConfig) {
+    this.isBlocked = config.isBlockedCallback;
+    this.topology = config.topology;
+  }
+
+  compute(initial: Vector2, goal: Vector2): Vector2[] | undefined {
+    //unpackage config.
+    const cameFrom = new Map<string, string>();
+    let processed: string[] = [];
+    let processing: Vector2[] = [];
+    processing.push(initial);
+
+    while (processing.length !== 0) {
+      //shift the last node off the array.
+      const current = processing.shift()!;
+      // check to see if the space is where you need to be. If so, exit loop.
+      if (Math.Vector.areEqual(current, goal)) {
+        let curStr = `${current.x}:${current.y}`;
+        const totalPath: Vector2[] = [current];
+
+        while (cameFrom.has(curStr)) {
+          //infinate loop here.
+          const prevStr = cameFrom.get(curStr)!;
+          const [x, y] = prevStr.split(":");
+          curStr = prevStr;
+          totalPath.unshift({
+            x: Number.parseInt(x),
+            y: Number.parseInt(y),
+          });
+        }
+        return totalPath;
+      }
+
+      //check adjactent spots to see if they are either blocked or don't exist.
+      let neighbors =
```

Comment:
Just hit an issue with this in AStar, think you'll hit the same thing here. Because getRing gets everything counter clockwise starting at like 3'oclock, the path can end up looking weird for topology = 8, even though it's correct.

For instance, if it wants to go straight up, it will sometimes go over to the right needlessly as it goes up, then come over to the left as it continues up, kind of like an arc. This is because the top-right neighbor gets evaluated before the top neighbor. While it is a shortest path, it can look weird. Check out how I added a new function in AStar just to compute neighbors a little more manually, and make it look more natural.